runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/WindmillStreamPool.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.dataflow.worker.windmill;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.function.Supplier;
+import javax.annotation.concurrent.ThreadSafe;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.annotations.VisibleForTesting;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+
+/**
+ * Pool of homogeneous streams to Windmill.
+ *
+ * <p>The pool holds a fixed total number of streams, and keeps each stream open for a specified
+ * time to allow for better load-balancing.
+ */
+@ThreadSafe
+public class WindmillStreamPool<StreamT extends WindmillStream> {
+
+  private final Duration streamTimeout;
+  private final Supplier<StreamT> streamSupplier;
+  private final Map<StreamT, StreamData<StreamT>> holds;
+  private final List<@Nullable StreamData<StreamT>> streamPool;
+
+  private WindmillStreamPool(
+      Duration streamTimeout,
+      Supplier<StreamT> streamSupplier,
+      List<@Nullable StreamData<StreamT>> streamPool,
+      Map<StreamT, StreamData<StreamT>> holds) {
+    this.streamPool = streamPool;
+    this.streamTimeout = streamTimeout;
+    this.streamSupplier = streamSupplier;
+    this.holds = holds;
+  }
+
+  public static <StreamT extends WindmillStream> WindmillStreamPool<StreamT> create(
+      int numStreams, Duration streamTimeout, Supplier<StreamT> streamSupplier) {
+    return new WindmillStreamPool<>(
+        streamTimeout, streamSupplier, newStreamList(numStreams), new ConcurrentHashMap<>());
```

Comment:
I don't think the ConcurrentHashMap is necessary because we are guarding holds by synchronizing on this.  Since it is more complex it is more overhead in performance and understanding so prefer to just use a hashmap.

You could add GuardedBy annotations to the guarded members.

Reply:
yea i realized since we are synchronizing other operations, we lose the benefits of concurrenthashmap.  changed to back to hashmap

done.